crates/server/src/store/blocklist/authority.rs

```diff
@@ -0,0 +1,430 @@
+// Copyright 2015-2022 Benjamin Fry <benjaminfry@me.com>
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// https://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// https://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+use std::{
+    collections::HashMap, fs::File, io, io::prelude::*, net::Ipv4Addr, path::Path, str::FromStr,
+};
+use tracing::{info, trace};
+
+#[cfg(feature = "dnssec")]
+use crate::{authority::Nsec3QueryInfo, config::dnssec::NxProofKind};
+use crate::{
+    authority::{
+        Authority, LookupControlFlow, LookupError, LookupObject, LookupOptions, MessageRequest,
+        UpdateResult, ZoneType,
+    },
+    proto::{
+        op::{Query, ResponseCode},
+        rr::{rdata::A, LowerName, Name, RData, Record, RecordType},
+    },
+    resolver::lookup::Lookup,
+    server::RequestInfo,
+    store::blocklist::{config::BlocklistConsultAction, BlocklistConfig},
+};
+
+/// A conditional authority that will resolve queries against one or more block lists and return a
+/// forged response.  The typical use case will be to use this in a chained configuration before a
+/// forwarding or recursive resolver in order to pre-emptively block queries for hosts that are on
+/// a block list. Refer to tests/test-data/test_configs/example_chained_recursor.toml for an example
+/// of this configuration.
+///
+/// The blocklist authority also supports the consult interface, which allows an authority to review
+/// a query/response that has been processed by another authority, and, optionally, overwrite that
+/// response before returning it to the requestor.  There is an example of this configuration in
+/// tests/test-data/test_configs/example_consulting_blocklist.toml.  The main intended use of this
+/// feature is to allow log-only configurations, to allow administrators to see if blocklist domains
+/// are being queried.  While this can be configured to overwrite responses, it is not recommended
+/// to do so - it is both more efficient, and more secure, to allow the blocklist to drop queries
+/// pre-emptively, as in the first example.
+///
+/// TODO:
+///  * Add (optional) support for logging the client IP address.  This will require some Authority
```

Comment:
In the current authority trait, request_info (which has the client IP) is passed to search, but not lookup.  This would be a change to add the client ip to lookup (and consult).

SIEM (Splunk, QRadar, etc.) users could probably link the server log with the client IP to the block log, but for grepability purposes (and for people who don't have a SIEM,) I'd like to have a single log line with the block action, query name, and client IP

Reply:
I'm ok with us passing that through I think, doesn't need to be this PR.

Reply:
Yep, I was going to make a separate PR for that change.